### PR TITLE
Optimized physics object spawn time

### LINKED ISSUE #30027: Bullet Physics is much slower then GodotPhysics
Godot version: Godot 3.1.1
OS/device including version: Windows 10 

Issue description:
When I'm adding 100,000 StaticBodies to the scene, each with CollisionShape, and set on each CollisionShape.disable to true, I have 1 FPS. But when I change to GodotPhysics, I have stable 60 FPS.
Event better, if I use GodotPhysics and set on each CollisionShape.disable to false, I still have 60 FPS (and when using Bullet - 1 FPS).

Steps to reproduce:
Do what I have wrote in Issue description

Here is a little sample project. Camera is not needed
[Link](http://www.lathuys.cba.pl/QuickSample.zip)

### PR DESCRIPTION
**Pull request by AndreaCatania**
- Optimized physics object spawn time
- Optimized shape usage when the shape is not scaled

fixes: #30027

The old bullet integration was 2 times slower compared to Godot physics when adding new bodies.
With this PR the timings got improved drastically (in the order of 1000 times).

(_All measurements were done with release debug_)

### Before optimization:
Bullet Physics Add 4000 Body time: 66 seconds
Godot Physics Add 4000 Body time: 28 seconds

### After optimization:
Bullet Physics Add 4000 Body time: ~0.07 seconds
Godot Physics Add 4000 Body time: 28 seconds

Bullet Physics Add 100000 Body time: ~6.5 seconds
Godot Physics Add 100000 Body time: **Crash after ~4500 bodies**

The tests got executed using:
```
extends Node

var shapes_per_frame := 4000
var shape := BoxShape3D.new()
var t1

func _ready():
	t1 = OS.get_unix_time()
	_do_test()


func _physics_process(delta):
	call_deferred("quit")


func quit():
	# Quit next frame.
	var t2 = OS.get_unix_time()
	print("Done in: ", t2 - t1)
	get_tree().quit()


func _do_test():
	print("Test start")
	for i in range(0, shapes_per_frame):
		var rb = RigidBody3D.new()
		var c = CollisionShape3D.new()
		c.set_shape(shape)
		rb.add_child(c)
		add_child(rb)
		if i % 50 == 0:
			print("Added: ", i)

```

### CODE REVIEW

**Comment by asheraryam** [comment]
I took a look at the changes and none of it seems to depend on 4.0 changes, so I believe this can apply to 3.2.x as well.

After this is merged I can port it to 3.2 if it would be welcome.

**Comment by aaronfranke** [comment]
This is a massive improvement, thank you for making this.

For testing, I [rebased this on the 3.2 branch](https://github.com/aaronfranke/godot/tree/3.2_add_body_impr_physics) and opened the [Voxel demo project](https://github.com/godotengine/godot-demo-projects/tree/master/3d/voxel) (but with VSync turned off). Each frame the demo generates 1 chunk. As the world generates, the FPS gradually becomes lower and lower.

* 3.2 branch: At render distance 6, min FPS is around 34. At render distance 10, min FPS is around 7.

* With [this commit on 3.2](https://github.com/aaronfranke/godot/tree/3.2_add_body_impr_physics): At render distance 6, min FPS is around 127. At render distance 10, min FPS is around 14.

So the performance in my case is significantly increased, and the gains are exponentially higher when there are fewer chunks in total (and therefore the slowness mostly comes from generating new ones rather than processing existing ones).

Also, the above is comparing a debug build with a debug build. With 3.2 release builds, at render distance 6, min FPS is around 120, and at render distance 10, min FPS is around 25. So my guess is that there will be similar % improvements with release builds, which would result in really high FPS (maybe ~400 min FPS at render distance 6 and ~50-60 at render distance 10).

File: modules/bullet/shape_bullet.cpp [hunk 1]
```diff
@@ -46,27 +46,56 @@
...
+	if (p_extra_edge == 0.0 && (p_implicit_scale - btVector3(1, 1, 1)).length2() <= CMP_EPSILON) {
+		return default_shape;
+	}
+
+	return internal_create_bt_shape(p_implicit_scale, p_extra_edge);
+}
+
+void ShapeBullet::destroy_bt_shape(btCollisionShape *p_shape) const {
+	if (p_shape != default_shape && p_shape != old_default_shape) {
+		bulletdelete(p_shape);
```
Comment (madmiraal) [inline modules/bullet/shape_bullet.cpp:73]: As far as I can tell, `old_default_shape` is always a `nullptr`.

File: modules/bullet/shape_bullet.cpp [hunk 2]
```diff
@@ -46,27 +46,56 @@
...
 		ShapeOwnerBullet *owner = static_cast<ShapeOwnerBullet *>(E->key());
 		owner->shape_changed(owner->find_shape(this));
 	}
+
+	if (old_default_shape) {
+		// At this point now one has the old default shape; just delete it.
+		bulletdelete(old_default_shape);
+		old_default_shape = nullptr;
+	}
 }
```
Comment (madmiraal) [inline modules/bullet/shape_bullet.cpp:99]: It appears as if `old_default_shape` is only used (and cleared) here; so I would just make it a local variable to this function.

File: modules/bullet/shape_bullet.h [hunk 3]
```diff
@@ -53,6 +53,10 @@ class ShapeBullet : public RIDBullet {
 	Map<ShapeOwnerBullet *, int> owners;
 	real_t margin = 0.04;
 
+	// Contains the default shape.
+	btCollisionShape *default_shape = nullptr;
+	btCollisionShape *old_default_shape = nullptr;
+
```
Comment (madmiraal) [inline modules/bullet/shape_bullet.h:59]: If `old_default_shape` is only used inside the one function, it does not need to be a class variable.

File: modules/bullet/space_bullet.cpp [hunk 4]
```diff
@@ -438,16 +440,27 @@ real_t SpaceBullet::get_param(PhysicsServer3D::SpaceParameter p_param) {
 }
 
 void SpaceBullet::add_area(AreaBullet *p_area) {
+#ifdef TOOLS_ENABLED
+	// This can't never happen at this piont.
+	CRASH_COND(p_area->is_in_world);
+#endif
```
Comment (madmiraal) [inline modules/bullet/space_bullet.cpp:446]: I'm not convinced that this can never happen. So, in case it does, instead of silently crashing with `CRASH_COND`, it would be better to use `ERR_FAIL_COND_MSG` with an explanation to assist future debugging and let the IDE continue running.

File: modules/bullet/space_bullet.cpp [hunk 5]
```diff
@@ -438,16 +440,27 @@ real_t SpaceBullet::get_param(PhysicsServer3D::SpaceParameter p_param) {
...
+	if (p_area->is_in_world) {
+		areas.erase(p_area);
+		dynamicsWorld->removeCollisionObject(p_area->get_bt_ghost());
+		p_area->is_in_world = false;
+	}
 }
 
 void SpaceBullet::reload_collision_filters(AreaBullet *p_area) {
+	if (p_area->is_in_world == false) {
+		return;
```
Comment (madmiraal) [inline modules/bullet/space_bullet.cpp:462]: It's a matter of style, but I prefer `if (!p_area->is_in_world)`, because I think it's more readable.

File: modules/bullet/space_bullet.cpp [hunk 6]
```diff
@@ -457,24 +470,43 @@ void SpaceBullet::reload_collision_filters(AreaBullet *p_area) {
...
+void SpaceBullet::unregister_collision_object(CollisionObjectBullet *p_object) {
+	collision_objects.erase(p_object);
+}
+
 void SpaceBullet::add_rigid_body(RigidBodyBullet *p_body) {
+#ifdef TOOLS_ENABLED
+	// This can't never happen at this piont.
+	CRASH_COND(p_body->is_in_world);
+#endif
 	if (p_body->is_static()) {
```
Comment (madmiraal) [inline modules/bullet/space_bullet.cpp:486]: Again, I'm not convinced this will never happen. We should use `ERR_FAIL_COND_MSG` with a good explanation, instead of `CRASH_COND`.

File: modules/bullet/space_bullet.cpp [hunk 7]
```diff
@@ -457,24 +470,43 @@ void SpaceBullet::reload_collision_filters(AreaBullet *p_area) {
...
+		} else {
+			dynamicsWorld->removeRigidBody(p_body->get_bt_rigid_body());
+		}
+		p_body->is_in_world = false;
 	}
 }
 
 void SpaceBullet::reload_collision_filters(RigidBodyBullet *p_body) {
+	if (p_body->is_in_world == false) {
+		return;
```
Comment (madmiraal) [inline modules/bullet/space_bullet.cpp:508]: Again, I prefer `if (!p_area->is_in_world)`.

File: modules/bullet/collision_object_bullet.h [hunk 8]
```diff
@@ -246,8 +262,12 @@ class RigidCollisionObjectBullet : public CollisionObjectBullet, public ShapeOwn
 	void set_shape_disabled(int p_index, bool p_disabled);
 	bool is_shape_disabled(int p_index);
 
+	virtual void prepare_object_for_dispatch();
+
 	virtual void shape_changed(int p_shape_index);
-	virtual void reload_shapes();
+	void reload_shapes();
+	bool need_reload_shapes() const { return need_shape_reload; }
+	virtual void do_reload_shapes();
```
Comment (madmiraal) [inline modules/bullet/collision_object_bullet.h:270]: As far a I can tell, the `need_reload_shapes()` function is never used.

File: modules/bullet/collision_object_bullet.h [hunk 9]
```diff
@@ -183,13 +190,21 @@ class CollisionObjectBullet : public RIDBullet {
 		return collisionLayer & p_other->collisionMask || p_other->collisionLayer & collisionMask;
 	}
 
-	virtual void reload_body() = 0;
+	bool need_reload_body() const {
+		return need_body_reload;
+	}
+
```
Comment (madmiraal) [inline modules/bullet/collision_object_bullet.h:196]: Again, `need_reload_body()` doesn't appear to be used.

File: modules/bullet/collision_object_bullet.cpp [hunk 10]
```diff
@@ -79,7 +79,7 @@ btTransform CollisionObjectBullet::ShapeWrapper::get_adjusted_transform() const
 }
 
 void CollisionObjectBullet::ShapeWrapper::claim_bt_shape(const btVector3 &body_scale) {
-	if (!bt_shape) {
+	if (bt_shape == nullptr) {
 		if (active) {
 			bt_shape = shape->create_bt_shape(scale * body_scale);
```
Comment (madmiraal) [inline modules/bullet/collision_object_bullet.cpp:84]: A matter of taste, but I prefer the original `if (!bt_shape)`, or better yet, the opposite: `if (bt_shape) return;` to avoid needing to indent the code.

File: modules/bullet/collision_object_bullet.cpp [hunk 11]
```diff
@@ -88,6 +88,13 @@ void CollisionObjectBullet::ShapeWrapper::claim_bt_shape(const btVector3 &body_s
 	}
 }
 
+void CollisionObjectBullet::ShapeWrapper::release_bt_shape() {
+	if (bt_shape != nullptr) {
+		shape->destroy_bt_shape(bt_shape);
```
Comment (madmiraal) [inline modules/bullet/collision_object_bullet.cpp:93]: Again, I prefer `if (!bt_shape) return;`

File: modules/bullet/collision_object_bullet.cpp [hunk 12]
```diff
@@ -319,58 +333,68 @@ bool RigidCollisionObjectBullet::is_shape_disabled(int p_index) {
...
-		shpWrapper = &shapes.write[0];
-		btTransform transform = shpWrapper->get_adjusted_transform();
+		// Is it possible to optimize by not using compound?
+		btTransform transform = shapes_ptr[0].get_adjusted_transform();
 		if (transform.getOrigin().isZero() && transform.getBasis() == transform.getBasis().getIdentity()) {
-			shpWrapper->claim_bt_shape(body_scale);
-			mainShape = shpWrapper->bt_shape;
+			shapes_ptr[0].claim_bt_shape(body_scale);
+			mainShape = shapes_ptr[0].bt_shape;
 			main_shape_changed();
```
Comment (madmiraal) [inline modules/bullet/collision_object_bullet.cpp:384]: Instead of `shapes_ptr[0].`, why not use `shapes_ptr->`?

**Review by madmiraal — COMMENTED** [review]
I've not actually tested the code, because it seems like many others have already, but aside from the relatively minor changes I've suggested, I think it looks good. :+1:

File: modules/bullet/space_bullet.cpp (hunk 7 above)
Comment (aaronfranke) [reply modules/bullet/space_bullet.cpp:508]: Yes, we should avoid `== false`, perhaps I can add this to the linter when I add it to the main Godot repository eventually.

File: modules/bullet/shape_bullet.cpp [hunk 13]
```diff
@@ -46,27 +46,56 @@
 	@author AndreaCatania
 */
 
-ShapeBullet::ShapeBullet() {}
+ShapeBullet::ShapeBullet() {
+}
 
-ShapeBullet::~ShapeBullet() {}
+ShapeBullet::~ShapeBullet() {
+	bulletdelete(default_shape);
```
[the remark this answers is not in the document]
Comment (AndreaCatania) [reply modules/bullet/shape_bullet.cpp:53]: Done

File: modules/bullet/shape_bullet.cpp [hunk 14]
```diff
@@ -46,27 +46,56 @@
...
+	default_shape = nullptr;
```
[the remark this answers is not in the document]
Comment (AndreaCatania) [reply modules/bullet/shape_bullet.cpp:54]: The reason I added it is for code style coherency. It doesn't cause any issue, so I prefer have it.

File: modules/bullet/shape_bullet.cpp (hunk 1 above)
Comment (AndreaCatania) [reply modules/bullet/shape_bullet.cpp:73]: During the shape refresh, done when a shape is changed, the `old_default_shape` contains the pointer of the old shape. This function is called in that phase to destroy the old shapes.

File: modules/bullet/shape_bullet.cpp (hunk 2 above)
Comment (AndreaCatania) [reply modules/bullet/shape_bullet.cpp:99]: The variable `old_default_shape` is used to avoid destroy the old default shape by the callbacks triggered by `owner->shape_changed(owner->find_shape(this));`

File: modules/bullet/shape_bullet.h (hunk 3 above)
Comment (AndreaCatania) [reply modules/bullet/shape_bullet.h:59]: The `old_default_shape` is used to not destroy the old default shape during shape refresh.

File: modules/bullet/space_bullet.cpp (hunk 4 above)
Comment (AndreaCatania) [reply modules/bullet/space_bullet.cpp:446]: This can't never happen, there is no way for the user to trigger it. It's there as notification bell for the future developers; if someone change the code and for a mistakes any body is added twice the condition is triggered and the debugger will tell him why.

By the way, bullet would crash anyway now that `btAssert` is again used.

File: modules/bullet/space_bullet.cpp (hunk 6 above)
Comment (AndreaCatania) [reply modules/bullet/space_bullet.cpp:486]: Same as above, I'm going to just enrich the comment.

File: modules/bullet/space_bullet.cpp (hunk 7 above)
Comment (AndreaCatania) [reply modules/bullet/space_bullet.cpp:508]: I find the explicit form `p_body->is_in_world == false` much more readable. However, if it's only matter of preference I leave it as is.

File: modules/bullet/collision_object_bullet.h (hunk 8 above)
Comment (AndreaCatania) [reply modules/bullet/collision_object_bullet.h:270]: Yes, but it's the getter of a variable, it's there for code completeness.

File: modules/bullet/collision_object_bullet.h (hunk 9 above)
Comment (AndreaCatania) [reply modules/bullet/collision_object_bullet.h:196]: Yes, but it's the getter of a variable, it's there for code completeness.

File: modules/bullet/collision_object_bullet.cpp (hunk 10 above)
Comment (AndreaCatania) [reply modules/bullet/collision_object_bullet.cpp:84]: The result is the same, but in this way it's much more clear that it's a pointer. The indentation just make the code more clear, in this case, for this reason I did it so.

File: modules/bullet/collision_object_bullet.cpp (hunk 12 above)
Comment (AndreaCatania) [reply modules/bullet/collision_object_bullet.cpp:384]: It's the pointer of an array, and just above is used as so. I'm not using the dereferencing syntax to keep the code coherent and more clear.

File: modules/bullet/shape_bullet.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (JohanAR) [reply modules/bullet/shape_bullet.cpp]: bulletdelete uses C++ delete so there is no point in checking for nullptr before calling it

**Comment by akien-mga** [comment]
Thanks!

I assume this might be worth backporting to the `3.2` branch, would you make a cherry-pick PR?

**Comment by AndreaCatania** [comment]
Yes, it would be a good backport. @asheraryam are you still up doing the back port?

File: modules/bullet/collision_object_bullet.cpp [hunk 15]
```diff
@@ -319,58 +333,68 @@ bool RigidCollisionObjectBullet::is_shape_disabled(int p_index) {
...
 	reload_shapes();
 }
 
 void RigidCollisionObjectBullet::reload_shapes() {
+	need_shape_reload = true;
+}
+
+void RigidCollisionObjectBullet::do_reload_shapes() {
 	if (mainShape && mainShape->isCompound()) {
 		// Destroy compound
```
Comment (madmiraal) [inline modules/bullet/collision_object_bullet.cpp:359]: Postponing the reloading of shapes until later causes crashes if those shapes are needed before `do_reload_shapes()` is run. #40311 and #40840 are just two simple examples. Even with additional triggers to call `do_reload_shapes()` (e.g. #40886 (or the [suggested alternative](https://github.com/godotengine/godot/pull/40886#issuecomment-666964762)) and #40252) the [crashes still happen](https://github.com/godotengine/godot/issues/40311#issuecomment-669153568).

`reload_shapes()` is called for numerous reasons:
- Creating a `RigidBody3D`
- Creating an `Area3D`
- Adding a `Shape`
- Setting a `Shape`
- Removing a `Shape`
- Changing a `Shape`
- Changing a `RigidBody's` size

Since each of these changes could be made before needing the shapes in a single user `_physics_process()` loop, there is no way to add sufficient calls to `do_reload_shape()` without effectively calling `do_reload_shape()` when calling `reload_shapes()`, which is exactly what this PR was attempting to avoid.

Furthermore, the test used to demonstrate the effectiveness of this PR is misleading, because it doesn't include the time for subsequently calling `do_reload_shapes()` on all the `RigidBodies` added. When including this time, the performance gains are eliminated.

Test project: [39726.zip](https://github.com/godotengine/godot/files/5028939/39726.zip)

Using this test project with 1,000 `RigidBodies` and `Shapes` on my laptop I get these total times (i.e. time since `_ready()` was called):

**Godot Physics:**
Frame 1: 3.470007
Frame 2: 36.684109
Frame 3: 36.751202
Frame 4: 36.815403
Frame 5: 36.876218
Frame 6: 36.935695
Frame 7: 36.99797
Frame 8: 37.061041
Frame 9: 37.289435
Frame 10: 37.353653


**Bullet Physics:**
Frame 1: 0.096981
Frame 2: 17.908696
Frame 3: 30.998907
Frame 4: 35.428883
Frame 5: 38.835653
Frame 6: 40.655049
Frame 7: 41.806311
Frame 8: 42.736726
Frame 9: 43.267167
Frame 10: 43.71381

In other words, not only does this PR cause numerous crashes that may never be resolvable, it doesn't achieve the performance gain it set out to achieve; the slow Bullet Frame Rate is still there.

Therefore, I think this PR should be reverted.

Comment (AndreaCatania) [reply modules/bullet/collision_object_bullet.cpp:359]: This is the code used for the test:
```GDScritp
extends Node

var count = 0

var shapes_per_frame := 1000
var shape := BoxShape3D.new()
var t1

func _ready():
	t1 = OS.get_unix_time()
	_do_test()

func _physics_process(delta):
	if count < 10:
		var t2 = OS.get_unix_time()
		count += 1
		print("Frame ", count, ": ", t2 - t1)

func _do_test():
	print("Test start")
	for i in range(0, shapes_per_frame):
		var rb = RigidBody3D.new()
		var c = CollisionShape3D.new()
		c.set_shape(shape)
		rb.add_child(c)
		add_child(rb)
		if i % 50 == 0:
			print("Added: ", i)
```

To be fair, I think that your test doesn't prove what you claim.

It adds 1000 `RigibBody` into the scene, into the `_ready` function. So to know how much time it taken to spawn those RigidBodies - you just need to check the first ever frame, that according to your measurements is slower in Godot Physics: `Frame 1: 3.470007` vs `Frame 1: 0.096981`. All other frame measurements are useless to prove spawn time and so the effectiveness of this PR is proven.

However, is important to say that this PR want to optimize the performance relative to Bullet itself and not relative to GodotPhysics which was running already well.

I don't think that the introduced bugs are a good reason to revert the optimization done instead of fixing the issues. It's like say that the Vulkan implementation is bad because is half broken, or the GDScript v2 is bad because it's not yet fully stable: and so both should be reverted.

The mentioned bug, especially #40840, have a solution: https://github.com/godotengine/godot/pull/40886#issuecomment-666964762 please let me know if you want to do that change, if no I can provide a new PR to fix that.

Comment (madmiraal) [reply modules/bullet/collision_object_bullet.cpp:359]: I'm not claiming anything. I'm refuting the claims made to support this PR; specifically that it solves #30027. I'm also highlighting why it has introduced numerous crashing issues, and suggesting that a PR that causes more problems than it solves should be reverted.

My test, which is a modified version of the original test provided to support this PR, shows that only looking at the time passed at the start of the first call to `_physics_process()` is misleading. It is misleading, because, at the time of the first call to `_physics_process()`, the calls to `do_reload_shapes()` have not occurred. They are called when the physics server's `step()` function is called, which happens after `_physics_process()`.

However, more importantly, this PR does not solve issue #30027. The PR reduces the number of times that the function that reloads the shapes is called, but these calls are not the source of the performance problem. The primary source of the performance problem is in the `step()` function; more specifically, calls to `btDiscreteDynamicsWorld::stepSimulation()` i.e. within Bullet physics.

While the actual gains of this PR are minimal, the issues caused by this change are significant. The more I investigate the crashing issues caused, the more I'm convinced they will not be solved with multiple band-aid fixes like [#40886 (comment)](https://github.com/godotengine/godot/pull/40886#issuecomment-666964762) without ultimately reverting this change, or worse: actually increasing the number of calls to `do_reload_shapes()`.

In summary, I believe this PR should be reverted, because it causes multiple crashing issues while doing little to improve actual performance.

Comment (AndreaCatania) [reply modules/bullet/collision_object_bullet.cpp:359]: Check this: https://github.com/godotengine/godot/pull/41082#issuecomment-670382832
